### PR TITLE
Stricter data type for dask keys

### LINKED ISSUE #10461: Stricter data type for dask keys
- Follow up to dask/distributed#8083 
- Blocks https://github.com/dask/distributed/issues/8130

dask/distributed#8083 recently redefined dask keys from any hashable to bytes, float, int, str, or tuples thereof.

While it is possible to run dask/dask on schedulers other than distributed, I don't think it's a healthy idea to have stricter restrictions in dask/distributed than in dask/dask.

# Actions
1. Create a new type alias,
```python
Key: TypeAlias = Union[str, bytes, int, float, tuple["Key", ...]]
```
2. Search the dask/dask codebase for all instances of `Hashable` and replace wherever applicable with `Key`
3. Update documentation, e.g. https://docs.dask.org/en/stable/spec.html#definitions
4. Move the new function `validate_key` from dask/distributed to dask/dask; add unit tests

### PR DESCRIPTION
**Pull request by crusaderky**
Closes #10461

### CODE REVIEW

File: dask/core.py [hunk 1]
```diff
@@ -193,32 +206,28 @@ def keys_in_tasks(keys, tasks, as_list=False):
...
-                    pass
-        tasks = work
-    return ret
+    typ = type(key)
+    if typ is tuple:
+        return all(iskey(i) for i in cast(tuple, key))
+    return typ in {bytes, int, float, str}
+
+
+def validate_key(key: object) -> None:
```
Comment (crusaderky) [inline dask/core.py:227]: Supersedes distributed.utils.validate_key

File: dask/base.py [hunk 2]
```diff
@@ -1518,10 +1518,20 @@ def get_name_from_key(key) -> str:
...
+def replace_name_in_key(key: str, rename: Mapping[str, str]) -> str:
+    ...
+
+
+@overload
+def replace_name_in_key(key: Key, rename: Mapping[str, str]) -> Key:
+    ...
+
+
+def replace_name_in_key(key, rename):
```
Comment (phofl) [inline dask/base.py:1534]: In pandas we would annotate this with the union, not sure if dask has a procedure here

File: dask/highlevelgraph.py [hunk 3]
```diff
@@ -148,14 +149,14 @@ def cull(
...
         return MaterializedLayer(out, annotations=self.annotations), ret_deps
 
-    def get_dependencies(self, key: Hashable, all_hlg_keys: Iterable) -> set:
+    def get_dependencies(self, key: Key, all_hlg_keys: Collection[Key]) -> set:
         """Get dependencies of `key` in the layer
 
         Parameters
         ----------
-        key: Hashable
+        key:
```
Comment (phofl) [inline dask/highlevelgraph.py:157]: Did you remove the type here on purpose?

**Review by phofl — APPROVED** [review]
very small non blocking comments, I don't have merge permissions though

File: dask/highlevelgraph.py (hunk 3 above)
Comment (crusaderky) [reply dask/highlevelgraph.py:157]: yes, there was an informal agreement that it is redundant with the type annotations. Happy to ridiscuss.

Comment (phofl) [reply dask/highlevelgraph.py:157]: We sometimes use more user-friendly annotations here, but I don't feel strongly about this at all

File: dask/base.py (hunk 2 above)
Comment (crusaderky) [reply dask/base.py]: The reason I didn't use the union is that the input determines the output and you really don't want to use unions as a return type (https://github.com/python/mypy/issues/1693). I've replaced the `@overload` with a more compact constrained `TypeVar` that achieves the same.

**Pull request merged**